tools/link.py

```diff
@@ -2377,80 +2377,101 @@ def node_pthread_detection():
 
-%(src)s
+%(generated_js)s
 
   return await moduleRtn;
 }
 ''' % {
-      'src': src,
+      'generated_js': generated_js
     }
   else:
-    src = '''
-%(maybe_async)sfunction(moduleArg = {}) {
+    wrapper_function = '''\
+%(maybe_async)sfunction %(wrapper_name)s(moduleArg = {}) {
   var moduleRtn;
 
-%(src)s
+%(generated_js)s
 
   return moduleRtn;
 }
 ''' % {
       'maybe_async': async_emit,
-      'src': src,
+      'generated_js': generated_js,
+      'wrapper_name': wrapper_name
     }
 
-  if settings.MINIMAL_RUNTIME and not settings.PTHREADS:
-    # Single threaded MINIMAL_RUNTIME programs do not need access to
-    # document.currentScript, so a simple export declaration is enough.
-    src = '/** @nocollapse */ var %s = %s' % (settings.EXPORT_NAME, src)
-  else:
+  if add_script_name_wrapper:
     script_url_node = ''
-    # When MODULARIZE this JS may be executed later,
-    # after document.currentScript is gone, so we save it.
-    # In EXPORT_ES6 + PTHREADS the 'thread' is actually an ES6 module
-    # webworker running in strict mode, so doesn't have access to 'document'.
-    # In this case use 'import.meta' instead.
-    if settings.EXPORT_ES6 and settings.USE_ES6_IMPORT_META:
-      script_url = 'import.meta.url'
-    else:
-      script_url = "typeof document != 'undefined' ? document.currentScript?.src : undefined"
-      if settings.ENVIRONMENT_MAY_BE_NODE:
-        script_url_node = "if (typeof __filename != 'undefined') _scriptName = _scriptName || __filename;"
+    # When MODULARIZE this JS may be executed later, after
+    # document.currentScript is gone, so we save it.  This is only
+    # needed if the program actually requires `_scriptName`
+    script_url = "typeof document != 'undefined' ? document.currentScript?.src : undefined"
+    if settings.ENVIRONMENT_MAY_BE_NODE:
+      script_url_node = "if (typeof __filename != 'undefined') _scriptName = _scriptName || __filename;"
     if settings.MODULARIZE == 'instance':
       src = '''\
   var _scriptName = %(script_url)s;
   %(script_url_node)s
-  %(src)s
+  %(wrapper_function)s
 ''' % {
         'script_url': script_url,
         'script_url_node': script_url_node,
-        'src': src,
+        'wrapper_function': wrapper_function,
       }
     else:
       src = '''\
 var %(EXPORT_NAME)s = (() => {
   var _scriptName = %(script_url)s;
   %(script_url_node)s
-  return (%(src)s);
+  return %(wrapper_function)s
 })();
 ''' % {
         'EXPORT_NAME': settings.EXPORT_NAME,
         'script_url': script_url,
         'script_url_node': script_url_node,
-        'src': src,
+        'wrapper_function': wrapper_function,
       }
+  elif add_assertion_wrapper:
+    src = '''\
+var %(EXPORT_NAME)s = (() => {
+  %(wrapper_function)s
+  // Return a small, never-async wrapper around %(wrapper_name)s which
+  // checks for callers using `new`.
+  return function(arg) {
+    if (new.target) throw new Error("%(EXPORT_NAME)s() must not be called with new");
```

Comment:
```suggestion
    if (new.target) throw new Error("%(EXPORT_NAME)s() must not be called as new %(EXPORT_NAME)s().");
```
I think otherwise the sentence might not be always understood by users (it could look clipped off, `X must not be called with new` - leading to "new what?" questions).